src/test/java/org/jenkinsci/plugins/gitclient/GitClientFetchTest.java

```diff
@@ -312,16 +306,21 @@ public void test_fetch_needs_preceding_prune() throws Exception {
         RefSpec defaultRefSpec = new RefSpec("+refs/heads/*:refs/remotes/origin/*");
         List<RefSpec> refSpecs = new ArrayList<>();
         refSpecs.add(defaultRefSpec);
-        try {
-            /* Fetch parent/a into newArea repo - fails for
-             * CliGitAPIImpl, succeeds for JGitAPIImpl */
-            newAreaWorkspace.launchCommand("git", "config", "fetch.prune", "false");
-            newAreaWorkspace.getGitClient().fetch(new URIish(newAreaWorkspace.getGitFileDir().toString()), refSpecs);
-            assertTrue("CliGit should have thrown an exception", newAreaWorkspace.getGitClient() instanceof JGitAPIImpl);
-        } catch (GitException ge) {
-            final String msg = ge.getMessage();
-            assertTrue("Wrong exception: " + msg, msg.contains("some local refs could not be updated") || msg.contains("error: cannot lock ref "));
-        }
+//        try {
+//            /* Fetch parent/a into newArea repo - fails for
+//             * CliGitAPIImpl, succeeds for JGitAPIImpl */
+//            newAreaWorkspace.launchCommand("git", "config", "fetch.prune", "false");
+//            newAreaWorkspace.getGitClient().fetch(new URIish(newAreaWorkspace.getGitFileDir().toString()), refSpecs);
+//            assertTrue("CliGit should have thrown an exception", newAreaWorkspace.getGitClient() instanceof JGitAPIImpl);
+//        } catch (GitException ge) {
+//            final String msg = ge.getMessage();
+//            assertTrue("Wrong exception: " + msg, msg.contains("some local refs could not be updated") || msg.contains("error: cannot lock ref "));
```

Comment:
I have left this code block commented because I was not sure if thrown.expectMessage is correctly implementing the assert condition here.

For now, I have coded two thrown.expectMessages instead of placing an || operator.

Reply:
Also, is it a good practice to use the same ExpectedException rule for two different types of errors in the same test class? Shouldn't we have a different test class for testing the other exception (Despite it is a GitException, it conveys a different message)?

Reply:
I see a misunderstanding in the use of expectedException in `test_fetch_needs_preceding_prune`.  I'm glad you raised the question, because I had not run the test in a debugger to confirm that it was reaching the end of the test.

The last executed statement of the test is the statement that is expected to throw the exception.  Any statements after the statement that throws the exception will not be executed.  In this case, the "Prune when a remote is not yet defined" declares that an exception is expected, then the `prune` call throws the exception.  All statements after that prune are ignored.

I will review at least some of the proposed changes to see if we can find a way to use the JUnit 4 facilities well and still test the paths that were tested before.  I suspect the change will be to create multiple tests from the original single overly large test.

Reply:
I just read the JUnit wiki for exception handling.  They note that [ExpectedException has been deprecated](https://github.com/junit-team/junit4/wiki/Exception-testing#expectedexception-rule) in JUnit 4.13.  My apologies that I had you learn a technique that is already deprecated.

I'll need to read that wiki page in more detail to understand the lambda based method that is preferred for asserting exceptions.

Reply:
I've submitted some small commits to [my copy of the branch](https://github.com/MarkEWaite/git-client-plugin/commits/JENKINS-60940) while I work through the test classes.

Thanks so much for the pull request!

Reply:
I've gone through your commits. All the discussion and changes is great for the development of the other test classes as well.

Would prefer a new PR with all the changes you've made or would you rather merge this and merge changes after that?

Reply:
I'll likely push my changes to your branch so that we continue evolving this pull request.  I'm struggling currently with a test that was transitioned and needed a significant change in the transition.

I believe that the `test_fetch_needs_preceding_prune` test was added fairly recently to GitAPITestCase, yet the updated copy in my working repo needs the key expected exception removed.

Reply:
https://github.com/MarkEWaite/git-client-plugin/blob/f5540088eb6f62718aff69d6e72e5e524f8eee77/src/test/java/org/jenkinsci/plugins/gitclient/GitClientFetchTest.java#L296-L306. I've seen that you have commented the part of code which asserts a throw. 

I don't understand it, if the client is an instance of CliGitAPIImpl, we expect it to throw a git exception?

Reply:
Yes, that's an unclear way of saying that JGit does not throw this exception but command line git does throw an exception.

Reply:
```newAreaWorkspace.getGitClient().fetch(new URIish(newAreaWorkspace.getGitFileDir().toString()), refSpecs);```

I think this is my mistake Mark, this line makes no sense. I am trying to fetch all branches from the same repository calling the fetch.

This should be 
```newAreaWorkspace.getGitClient().fetch(new URIish(bareWorkspace.getGitFileDir().toString()), refSpecs);``` 
So we are trying to fetch parent/a from the bare repository.

Reply:
Ah, I'm so glad you detected that!  I failed to look at the source that was being used for the clone!  Thanks very much.